### PR TITLE
include mkl libraries for intel on linux and MacOS

### PR DESCRIPTION
**Pull request by iulusoy**
- mkl libraries included for most MacOS versions
- mkl libraries included for all linux versions
- new test for mkl installation
- currently is not the default, the mkl test is run conditionally. With the inclusion of math libraries for more compilers+OS, the tests should run as default

### CODE REVIEW

**Review by wpbonelli — COMMENTED** [review]
to get ci tests to run you can add the math branch to the `pull_request` trigger in `test.yml`

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -85,6 +87,17 @@ jobs:
           compiler: ${{ matrix.toolchain.compiler }}
           version: ${{ matrix.toolchain.version }}
       
+      - name: Test MKL libraries
+        # we could also exclude 2021.5 for all macos versions here
+        # this needs refactoring at some point - should the install
+        # fail if no mkl version exists?
```
Comment (wpbonelli) [inline .github/workflows/test.yml:93]: I think it makes sense to fail, given we aim to guarantee full support barring a few exceptions. In general I think we should fail if we can't satisfy the user's selected configuration.

**Comment by wpbonelli** [comment]
@iulusoy I rebased `math` from `main`, could you rebase this or enable contributions? I think we could merge soon and keep iterating

**Comment by wpbonelli** [comment]
thanks @iulusoy I joined. in future you can just check this box to give a repo maintainer access to a PR branch from your fork without making them a full contributor

allow_edits

File: README.md [hunk 2]
```diff
@@ -35,6 +35,7 @@ jobs:
           - {compiler: gcc, version: 13}
           - {compiler: intel, version: '2023.2'}
           - {compiler: intel-classic, version: '2021.10'}
+          - {compiler: intel-classic, version: '2021.10'}
```
Comment (wpbonelli) [inline README.md:38]: duplicates line above

File: README.md [hunk 3]
```diff
@@ -48,11 +49,13 @@ jobs:
             toolchain: {compiler: nvidia-hpc, version: '23.11'}
 
     steps:
-      - uses: fortran-lang/setup-fortran@v1
+      - uses: ssciwr/setup-fortran@v0.1
```
Comment (wpbonelli) [inline README.md:52]: this should stay `fortran-lang/setup-fortran@v1`

**Comment by wpbonelli** [comment]
@iulusoy can you revert the CSV changes and/or rebase? I updated `math` from `main`.

In general you can ignore the auto-update PRs on your fork (really they should just run on the main repo, I'll change that at some point) and just rebase from upstream instead.

**Comment by wpbonelli** [comment]
@iulusoy any idea on the test failures?

**Comment by iulusoy** [comment]
> @iulusoy any idea on the test failures?

No, not really - I need to try this out in a container to see what is happening. Have it on my to-do list.

What about extending the math support to other compilers/windows? Should we assign responsibilities to move this forward? 
I will have some more time for this towards the end of next month, so I could look at gnu compiler; but I have no experience with installing them on windows and would like to pass that to someone who is more experienced with it.

**Comment by wpbonelli** [comment]
> What about extending the math support to other compilers/windows? Should we assign responsibilities to move this forward?
I will have some more time for this towards the end of next month, so I could look at gnu compiler; but I have no experience with installing them on windows and would like to pass that to someone who is more experienced with it.

I'm still not sure on the approach for other toolchains. Since [BLAS/LAPACK were added to `stdlib`](https://github.com/fortran-lang/setup-fortran/issues/55#issuecomment-2052696763) an alternative besides those mentioned in #55 could be an `install_stdlib` option, which could install the reference implementation, and `install_mkl` could be a one-off specific to the Intel toolchain.

This PR could be merged with mac/linux support and windows could come in a separate PR, then `math` could merge into `main` and we could release this as an Intel-specific feature.

Just thinking out loud

**Comment by iulusoy** [comment]
> @iulusoy any idea on the test failures?

Seems that the apt installation of the mkl libraries changes the setvars script for mkl version 2021.1. Because of this I changed the order of the compiler vs mkl installation. The PR will be updated shortly.

**Comment by iulusoy** [comment]
@wpbonelli do you think this can be merged now, or do we need some more tests?

**Comment by wpbonelli** [comment]
@iulusoy do you mind doing a final rebase? looks like the diff still has changes in the compat scripts and CSVs.

**Comment by iulusoy** [comment]
> @iulusoy do you mind doing a final rebase? looks like the diff still has changes in the compat scripts and CSVs.

Should be no more diffs to main now, other than the actual changes.

**Comment by wpbonelli** [comment]
Looks like this is still behind. Maybe helpful to squash everything down to your first commit e.g. `git reset --soft 56f36a1`, `git commit --amend --no-edit` before doing an interactive rebase so you only have to resolve conflicts once.

**Review by wpbonelli — APPROVED** [review]
looks good to me, merging now.

**Pull request merged**